pkg/header/check.go

```diff
@@ -28,92 +30,144 @@ import (
+	"github.com/go-git/go-billy/v5/osfs"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/format/gitignore"
+	"github.com/go-git/go-git/v5/plumbing/object"
 )
 
 // Check checks the license headers of the specified paths/globs.
 func Check(config *ConfigHeader, result *Result) error {
-	for _, pattern := range config.Paths {
-		if err := checkPattern(pattern, result, config); err != nil {
+	fileList, err := listFiles(config)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range fileList {
+		if err := CheckFile(file, config, result); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
-var seen = make(map[string]bool)
+func listFiles(config *ConfigHeader) ([]string, error) {
+	var fileList []string
 
-func checkPattern(pattern string, result *Result, config *ConfigHeader) error {
-	paths, err := glob(pattern)
+	repo, err := git.PlainOpen("./")
 
-	if err != nil {
-		return err
-	}
+	if err != nil { // we're not in a Git workspace, fallback to glob paths
+		var localFileList []string
+		for _, pattern := range config.Paths {
+			if pattern == "." {
+				pattern = "./"
+			}
+			files, err := doublestar.Glob(pattern)
+			if err != nil {
+				return nil, err
+			}
+			localFileList = append(localFileList, files...)
+		}
+
+		seen := make(map[string]bool)
+		for _, file := range localFileList {
+			files, err := walkFile(file, seen)
+			if err != nil {
+				return nil, err
+			}
+			fileList = append(fileList, files...)
+		}
+	} else {
+		var candidates []string
 
-	for _, path := range paths {
-		if yes, err := config.ShouldIgnore(path); yes || err != nil {
-			result.Ignore(path)
-			continue
+		t, _ := repo.Worktree()
+		if t.Excludes == nil {
+			t.Excludes = make([]gitignore.Pattern, 0)
 		}
-		if err := checkPath(path, result, config); err != nil {
-			return err
+		if ignorePattens, err := gitignore.LoadGlobalPatterns(osfs.New("")); err == nil {
+			t.Excludes = append(t.Excludes, ignorePattens...)
+		}
+		if ignorePattens, err := gitignore.LoadSystemPatterns(osfs.New("")); err == nil {
+			t.Excludes = append(t.Excludes, ignorePattens...)
+		}
+		s, _ := t.Status()
+		for file := range s {
+			candidates = append(candidates, file)
 		}
-		seen[path] = true
-	}
 
-	return nil
-}
+		head, _ := repo.Head()
+		commit, _ := repo.CommitObject(head.Hash())
+		tree, err := commit.Tree()
+		if err != nil {
+			return nil, err
+		}
+		if err := tree.Files().ForEach(func(file *object.File) error {
+			if file == nil {
+				return errors.New("file pointer is nil")
+			}
+			if _, err := os.Stat(file.Name); err == nil {
+				candidates = append(candidates, file.Name)
+			}
+			return nil
+		}); err != nil {
+			return nil, err
+		}
 
-func glob(pattern string) (matches []string, err error) {
-	if pattern == "." {
-		return doublestar.Glob("./")
+		seen := make(map[string]bool)
+		for _, file := range candidates {
+			if !seen[file] {
+				seen[file] = true
+				if _, err := os.Stat(file); err == nil {
```

Comment:
Same here

Reply:
See https://github.com/apache/skywalking-eyes/pull/133/files#r962437870. Although, we can merge these two blocks.